### DISCUSSION TITLE
Some questions for help. 1. No constants are introduced. 2. Mutation of variables. 3. A bug of custom loss function.

### DISCUSSION THREAD

**Discussion by djw-easy** [MilesCranmer/PySR#968] [Q&A]
Hi Miles, 

I have encountered some problems while using PySR and hope to get your reply.

# 1. No constants are introduced

I don't want constants in my expressions, so I set it up as follows, but I'm not sure if it's correct. Also, I'm not sure if this setting will affect the search efficiency of PySR.

```python
model = PySRRegressor(
  maxsize=30, maxdepth=30, 
  binary_operators=["+", "*", "-", "/"],
  complexity_of_constants=100
)
```

# 2. Mutation of variables

I noticed that the mutation weights of PySR are only set for operators and constants. Is it possible to mutate variables, for example, `y=2*x0` mutates to `y=2*x1`? This will be very helpful for the convergence of the expression of my research.

# 3. A bug of custom loss function

Below is the loss function I use, which outputs a loss value between 0 and 1. In order to prevent the output expression from being a single value (such as a constant), I added an equality check, which returns 1. However, I found that this equality check sometimes does not work in actual operation, such as the expression `y = ((((x0 - x1) + x1) - x0) / x1) / x2`, in which case the loss value is 0.

```python
hellinger_loss = """
function hellinger_loss(tree, dataset::Dataset{T,L}, options)::L where {T,L}
    prediction, flag = eval_tree_array(tree, dataset.X, options)
    if !flag
        return L(Inf) # Return Inf if evaluation failed
    end

    # Use approximate equality check for floating point numbers
    eps_tol = L(1e-6)
    pred_range = maximum(prediction) - minimum(prediction)
    
    # Check if all predictions are approximately the same
    if pred_range < eps_tol
        return L(1.0) # If all predictions are the same, return 1 loss
    end

    Omit some code
    return loss
end
"""
```
